pkg/container-collection/container-collection.go

```diff
@@ -0,0 +1,228 @@
+// Copyright 2019-2021 The Inspektor Gadget authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package containercollection provides the ContainerCollection struct to keep
+// track of the set of running containers and primitives to query that set with
+// various criteria.
+//
+// It is used by the Gadget Tracer Manager to keep track of containers part of
+// Kubernetes pods and by Local Gadget Manager to keep track of containers on a
+// Linux system.
+package containercollection
+
+import (
+	"sync"
+
+	pb "github.com/kinvolk/inspektor-gadget/pkg/gadgettracermanager/api"
+	"github.com/kinvolk/inspektor-gadget/pkg/gadgettracermanager/pubsub"
+)
+
+// ContainerCollection holds a set of containers. It can be embedded as an
+// anonymous struct to help other structs implement the ContainerResolver
+// interface. For this reason, some methods are prefixed with 'Container' to
+// make clear that
+
+type ContainerCollection struct {
+	// Keys:   Container ID
+	// Values: pb.ContainerDefinition
+	containers sync.Map
+
+	// subs contains a list of subscribers of container events
+	pubsub *pubsub.GadgetPubSub
+}
+
+// ContainerCollectionInitialize initializes a ContainerCollection. It is
+// useful when ContainerCollection is embedded as an anonymous struct because
+// we don't use a contructor in that case.
+func (cc *ContainerCollection) ContainerCollectionInitialize() {
+	cc.pubsub = pubsub.NewGadgetPubSub()
+}
+
+// ContainerGet looks up a container by the container id and return it if
+// found. The ok result indicates whether value was found in the map.
+func (cc *ContainerCollection) ContainerGet(id string) (value *pb.ContainerDefinition, ok bool) {
+	v, ok := cc.containers.Load(id)
+	if !ok {
+		return nil, false
+	}
+	containerDefinition := v.(pb.ContainerDefinition)
+	return &containerDefinition, true
+}
+
+// ContainerRemove removes a container from the collection.
+func (cc *ContainerCollection) ContainerRemove(id string) {
+	c, ok := cc.ContainerGet(id)
+	if !ok {
+		return
+	}
+	cc.pubsub.Publish(pubsub.EVENT_TYPE_REMOVE_CONTAINER, *c)
+	cc.containers.Delete(id)
+}
+
+// ContainerAdd adds a container to the collection.
+func (cc *ContainerCollection) ContainerAdd(container pb.ContainerDefinition) {
+	cc.containers.Store(container.Id, container)
+	cc.pubsub.Publish(pubsub.EVENT_TYPE_ADD_CONTAINER, container)
+}
```

Comment:
I'd like to get others' opinion but I think `AddContainer` is a much better name. @marga-kinvolk @blanquicet

Reply:
I am fine changing it. I named it like that to have the `Container` prefix (see godoc of `type ContainerCollection struct`).

Reply:
I like the idea of having the `Container` prefix but I agree that `AddContainer` sounds better. In addition, we have used that notation so far and it would be strange to have `GadgetTracerManager.AddContainer` and `LocalGadgetManager.AddContainer` calling `ContainerCollection.ContainerAdd`, don't you think?

Reply:
Ok, I will name the methods as follow:
* `LocalGadgetManager.addContainerRuncNotif` (see https://github.com/kinvolk/inspektor-gadget/pull/245#discussion_r712367224)
* `ContainerCollection.AddContainer`

Reply:
Ah, naming `ContainerCollection.AddContainer` does not work with the Gadget Tracer Manager because the name conflict with `GadgetTracerManager.AddContainer` and the latter cannot be changed because it comes from the gRPC proto.

Since `ContainerCollection` is present as anonymous fields in both `GadgetTracerManager` and `LocalTracerManager`, its methods are promoted, so it does not compile with the name conflict. Using the method promotions is necessary so that both `GadgetTracerManager` and `LocalTracerManager` implement the Resolver interface for the gadgets to use methods such as `LookupMntnsByPod`.

Reply:
Oh, I didn't realise that anonymous fields can still be named like this:
```
g.ContainerCollection.AddContainer(*containerDefinition)
```
I'll try that.